pkg/ingester/owned_series.go

```diff
@@ -87,13 +93,23 @@ func newOwnedSeriesService(
 		}),
 	}
 
-	oss.Service = services.NewTimerService(interval, nil, oss.onPeriodicCheck, nil)
+	oss.Service = services.NewTimerService(interval, oss.starting, oss.checkOwnedSeries, nil)
 	return oss
 }
 
+func (oss *ownedSeriesService) starting(ctx context.Context) error {
+	err := oss.ringStrategy.waitForRing(ctx)
+	if err != nil {
+		return fmt.Errorf("initial wait for ring: %v", err)
+	}
+
+	_ = oss.checkOwnedSeries(ctx)
```

Comment:
You're right. Let me make this work only for ingester ring case. 

> refactoring the ingester starting method to start the owned series service before the lifecycler?

For partitions ring this doesn't work -- if there is no partition entry, owned series would never start. Then partition lifecycler would also not start, and but we need it to create the partition entry.

Reply:
> Let me make this work only for ingester ring case.

Done in https://github.com/grafana/mimir/pull/7583/commits/24eefbd190c49266884bd0fb70d0556aa9c3a4a5

Reply:
> There is actually a remaining bit of refactoring for the owned series service to address https://github.com/grafana/mimir/pull/7087#discussion_r1455032145. @pracucci and I had discussed moving the initial check logic back into the owned series service starting method, but not using the service manager to start it (instead starting it separately, [like the lifecycler is started](https://github.com/grafana/mimir/blob/f9d168c67f1eb7181232d5ae6de7d02e16e1940b/pkg/ingester/ingester.go#L550-L556)).

I've implemented this solution now. `ownedSeriesService` is started asynchronously, and performs initial check in the `Starting` state, but ingester doesn't wait until `ownedSeriesService` is `Running`, and proceeds. This unblocks reads. However pushes are blocked until `ownedSeriesService` enters `Running` state (when limiting based on owned series is enabled), to make sure that pushes work with correct limits.

Reply:
Thinking more about this I have a couple of concerns:

1) The ingester doesn't wait for the owned series service to become `Running`, and we don't have any kind of limit on how long it will wait for the ring in the `Starting` state. However unlikely, isn't it now possible for the ingester to run forever accepting reads but rejecting writes because the owned series service is stuck in `Starting`? Should we have a failsafe where we stop the ingester if the owned series service doesn't start in a timely fashion?

2) Because we allow the ingester to become ACTIVE in the ring and then reject writes until the owned series service is `Running`, I worry about a smaller-scale version of the issue described [here](https://github.com/grafana/mimir/pull/7087) happening, this time on the write path.

Unlike for reads, we don't need the full ingester zones to be healthy to have quorum for writes. However, we still require RF/2 + 1 successes. In the case of scaling up ingesters, we add ingesters to all zones at once. This means that it is possible for the new ingesters in all zones to be in a state where they are ACTIVE in the ring but rejecting writes waiting for the owned series service. If a distributor tries to write series to multiple ingesters in this state, we'll fail the write.

A similar situation can happen during rollouts -- `zone 1` ingesters are restarted, move to `ACTIVE`, and reject writes while the owned series are calculated; `zone 2` ingesters are restarted because the rollout operator sees `zone 1` as healthy, and writes start to fail because two zones are unavailable/failing.

These write failures should be retried, but I feel like we should avoid these errors if we can.

---

I'm wondering if there's really a downside to running the owned series computation regardless of whether the ingester/partition is in the ring or not. In the "good" case, the ingester/partition is there and we do the correct calculation. In the "bad" case, the entry is missing from the ring, and we'll use `nil` token ranges, and set the owned series count to `0`. This would be fixed the next time the owned series service runs, since the ring change will be detected once the ingester/partition is actually added to the ring. The downside is that we potentially accept writes that should be rejected due to limits.

The "worst" case is when we have an empty ring. In this case we can't wait for the ring to exist because that would mean waiting for the lifecycler again. BUT, if we're in this state, either a) we're building the cluster for the first time, in which case all the series counts should be 0 anyway, or b) the ring state was completely destroyed and we're recovering from a disaster, in which case maybe it's ok to run with the incorrect owned series counts until the next owned series service iteration ([here](https://github.com/grafana/mimir/pull/7087#issue-2073630536) I say that in this case the owned series counts will be 0, but now I think I was wrong).

Thoughts?

Reply:
> The ingester doesn't wait for the owned series service to become Running, and we don't have any kind of limit on how long it will wait for the ring in the Starting state.

You're right. I reasoned that this will be very unlikely state, but the risk exists.

> I'm wondering if there's really a downside to running the owned series computation regardless of whether the ingester/partition is in the ring or not.

I am fine with this approach. If there's ring is empty or entry doesn't exist, we can run the computation, but perform checks more often until entry actually *does* exist. (The checks will be no-op until entry appears in the ring).

> The downside is that we potentially accept writes that should be rejected due to limits.

Right, this worries me a bit, but we can minimize the problem by running checks more often when we detect that initial check was performed with empty ring or without our instance.